File: ios/ClovisMobile/Sources/Screens/DossiersScreen.swift
// Cree le 23/08/2026, Bourama : Lot 2 Partie 3 (app mobile), fichiers/dossiers designes, iOS.
//
// Meme principe que DossiersScreen.kt cote Android : deux niveaux (liste des
// dossiers designes, puis navigation a l'interieur d'un dossier via une pile
// en memoire). UIDocumentPickerViewController n'a pas d'equivalent SwiftUI
// natif -- on le passe par UIViewControllerRepresentable.
import SwiftUI
import UniformTypeIdentifiers

struct DossiersScreen: View {
    @State private var dossiersDesignes: [DossierDesigne] = DossiersDesignesRepository.listerDossiersDesignes()
    @State private var pile: [(url: URL, nom: String)] = []
    @State private var contenu: [ElementDossier] = []
    @State private var afficherSelecteur = false
    @State private var afficherNouveauDossier = false
    @State private var afficherNouveauFichier = false
    @State private var elementARenommer: ElementDossier?
    @State private var elementASupprimer: ElementDossier?
    @State private var nomSaisi = ""
    @State private var erreur: String?

    private func rafraichirContenu() {
        guard let courant = pile.last else { return }
        contenu = DossiersDesignesRepository.listerContenu(courant.url)
    }

    var body: some View {
        Group {
            if let courant = pile.last {
                vueContenuDossier(courant: courant)
            } else {
                vueRacine
            }
        }
        .sheet(isPresented: $afficherSelecteur) {
            SelecteurDossier { url in
                DossiersDesignesRepository.ajouterDossierDesigne(url: url)
                dossiersDesignes = DossiersDesignesRepository.listerDossiersDesignes()
            }
        }
    }

    private var vueRacine: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("Dossiers désignés").font(.title2.bold())
            Text("Clovis ne peut lire ou modifier que les dossiers que tu choisis explicitement ici.")
                .font(.subheadline)
                .foregroundColor(.secondary)

            Button {
                afficherSelecteur = true
            } label: {
                Label("Désigner un dossier", systemImage: "plus")
                    .frame(maxWidth: .infinity)
            }
            .buttonStyle(.borderedProminent)

            if dossiersDesignes.isEmpty {
                Text("Aucun dossier désigné pour l'instant.")
                    .foregroundColor(.secondary)
            }

            List {
                ForEach(dossiersDesignes) { dossier in
                    HStack {
                        Image(systemName: "folder")
                        Text(dossier.url.lastPathComponent)
                        Spacer()
                        Button {
                            DossiersDesignesRepository.retirerDossierDesigne(id: dossier.id)
                            dossiersDesignes = DossiersDesignesRepository.listerDossiersDesignes()
                        } label: {
                            Image(systemName: "xmark.circle")
                        }
                        .buttonStyle(.plain)
                    }
                    .contentShape(Rectangle())
                    .onTapGesture {
                        pile = [(url: dossier.url, nom: dossier.url.lastPathComponent)]
                        rafraichirContenu()
                    }
                }
            }
            .listStyle(.plain)
        }
        .padding(24)
    }

    private func vueContenuDossier(courant: (url: URL, nom: String)) -> some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack {
                Button {
                    pile.removeLast()
                    if !pile.isEmpty { rafraichirContenu() }
                } label: {
                    Image(systemName: "chevron.left")
                }
                Text(pile.map { $0.nom }.joined(separator: " / "))
                    .font(.headline)
                    .lineLimit(1)
            }

            HStack(spacing: 12) {
                Button {
                    nomSaisi = ""
                    afficherNouveauDossier = true
                } label: {
                    Label("Dossier", systemImage: "folder.badge.plus")
                }
                Button {
                    nomSaisi = ""
                    afficherNouveauFichier = true
                } label: {
                    Label("Fichier", systemImage: "doc.badge.plus")
                }
            }
            .buttonStyle(.bordered)

            if let erreur {
                Text(erreur).foregroundColor(.red).font(.footnote)
            }

            if contenu.isEmpty {
                Text("Dossier vide.").foregroundColor(.secondary)
            }

            List {
                ForEach(contenu) { element in
                    HStack {
                        Image(systemName: element.estDossier ? "folder" : "doc")
                        VStack(alignment: .leading) {
                            Text(element.nom)
                            if !element.estDossier {
                                Text("\(element.tailleOctets / 1024) Ko")
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                            }
                        }
                        Spacer()
                        Button {
                            elementARenommer = element
                            nomSaisi = element.nom
                        } label: {
                            Image(systemName: "pencil")
                        }
                        .buttonStyle(.plain)
                        Button {
                            elementASupprimer = element
                        } label: {
                            Image(systemName: "trash")
                        }
                        .buttonStyle(.plain)
                    }
                    .contentShape(Rectangle())
                    .onTapGesture {
                        if element.estDossier {
                            pile.append((url: element.url, nom: element.nom))
                            rafraichirContenu()
                        }
                    }
                }
            }
            .listStyle(.plain)
        }
        .padding(24)
        .alert("Nouveau dossier", isPresented: $afficherNouveauDossier) {
            TextField("Nom du dossier", text: $nomSaisi)
            Button("Annuler", role: .cancel) {}
            Button("Créer") {
                erreur = DossiersDesignesRepository.creerSousDossier(dansParent: courant.url, nom: nomSaisi) ? nil : "Impossible de créer ce dossier."
                rafraichirContenu()
            }
        }
        .alert("Nouveau fichier", isPresented: $afficherNouveauFichier) {
            TextField("Nom du fichier (avec extension)", text: $nomSaisi)
            Button("Annuler", role: .cancel) {}
            Button("Créer") {
                erreur = DossiersDesignesRepository.creerFichier(dansParent: courant.url, nom: nomSaisi) ? nil : "Impossible de créer ce fichier."
                rafraichirContenu()
            }
        }
        .alert("Renommer", isPresented: Binding(get: { elementARenommer != nil }, set: { if !$0 { elementARenommer = nil } })) {
            TextField("Nouveau nom", text: $nomSaisi)
            Button("Annuler", role: .cancel) { elementARenommer = nil }
            Button("Renommer") {
                if let element = elementARenommer {
                    erreur = DossiersDesignesRepository.renommer(element.url, nouveauNom: nomSaisi) ? nil : "Impossible de renommer."
                    rafraichirContenu()
                }
                elementARenommer = nil
            }
        }
        .alert(
            elementASupprimer.map { "Supprimer « \($0.nom) » ?" } ?? "",
            isPresented: Binding(get: { elementASupprimer != nil }, set: { if !$0 { elementASupprimer = nil } })
        ) {
            Button("Annuler", role: .cancel) { elementASupprimer = nil }
            Button("Supprimer", role: .destructive) {
                if let element = elementASupprimer {
                    erreur = DossiersDesignesRepository.supprimer(element.url) ? nil : "Impossible de supprimer."
                    rafraichirContenu()
                }
                elementASupprimer = nil
            }
        } message: {
            Text(elementASupprimer?.estDossier == true ? "Le dossier et tout son contenu seront supprimés." : "Le fichier sera supprimé définitivement.")
        }
    }
}

/// Wrapper UIKit -> SwiftUI pour UIDocumentPickerViewController en mode
/// selection de dossier (SwiftUI n'a pas d'equivalent natif a ce jour).
private struct SelecteurDossier: UIViewControllerRepresentable {
    var onChoisi: (URL) -> Void

    func makeUIViewController(context: Context) -> UIDocumentPickerViewController {
        let picker = UIDocumentPickerViewController(forOpeningContentTypes: [.folder])
        picker.delegate = context.coordinator
        return picker
    }

    func updateUIViewController(_ uiViewController: UIDocumentPickerViewController, context: Context) {}

    func makeCoordinator() -> Coordinator { Coordinator(onChoisi: onChoisi) }

    class Coordinator: NSObject, UIDocumentPickerDelegate {
        let onChoisi: (URL) -> Void
        init(onChoisi: @escaping (URL) -> Void) { self.onChoisi = onChoisi }

        func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
            if let url = urls.first { onChoisi(url) }
        }
    }
}

File: android-legacy-natif/app/src/main/java/com/clovis/app/ui/screens/DossiersScreen.kt
// Cree le 23/08/2026, Bourama : Lot 2 Partie 3 (app mobile), fichiers/dossiers designes.
//
// Deux niveaux : liste des dossiers designes (racine de l'ecran), puis
// navigation a l'interieur d'un dossier (pile de DossierDesigne -> sous-
// dossiers, style breadcrumb). Pas de bibliotheque de navigation ajoutee :
// meme approche state-driven que MainActivity.kt (Lot 1), une simple pile
// en memoire suffit pour ce niveau de profondeur.
package com.clovis.app.ui.screens

import android.net.Uri
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import com.clovis.app.data.DossierDesigne
import com.clovis.app.data.DossiersDesignesRepository
import com.clovis.app.data.ElementDossier

@Composable
fun DossiersScreen() {
    val context = LocalContext.current
    val repo = remember { DossiersDesignesRepository(context) }

    var dossiersDesignes by remember { mutableStateOf(repo.listerDossiersDesignes()) }
    // Pile de navigation : premier element = dossier designe ouvert, suivants = sous-dossiers.
    // Chaque entree : (uri du dossier courant, nom affiche, uri du PARENT direct pour un eventuel deplacement).
    var pile by remember { mutableStateOf<List<Triple<Uri, String, Uri?>>>(emptyList()) }
    var contenu by remember { mutableStateOf<List<ElementDossier>>(emptyList()) }

    var dialogueNouveauDossier by remember { mutableStateOf(false) }
    var dialogueNouveauFichier by remember { mutableStateOf(false) }
    var elementARenommer by remember { mutableStateOf<ElementDossier?>(null) }
    var elementASupprimer by remember { mutableStateOf<ElementDossier?>(null) }
    var erreur by remember { mutableStateOf<String?>(null) }

    fun rafraichirContenu() {
        val courant = pile.lastOrNull() ?: return
        contenu = repo.listerContenu(courant.first)
    }

    val selecteurDossier = rememberLauncherForActivityResult(
        ActivityResultContracts.OpenDocumentTree()
    ) { uri ->
        if (uri != null) {
            repo.ajouterDossierDesigne(uri)
            dossiersDesignes = repo.listerDossiersDesignes()
        }
    }

    // --- Vue "a l'interieur d'un dossier" ---
    if (pile.isNotEmpty()) {
        val courant = pile.last()

        LaunchedEffect(courant.first) { rafraichirContenu() }

        Column(modifier = Modifier.fillMaxSize().padding(20.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                IconButton(onClick = { pile = pile.dropLast(1) }) {
                    Icon(Icons.Default.ArrowBack, contentDescription = "Retour")
                }
                Text(
                    pile.joinToString(" / ") { it.second },
                    style = MaterialTheme.typography.titleMedium,
                    modifier = Modifier.weight(1f)
                )
            }
            Spacer(Modifier.height(8.dp))

            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                OutlinedButton(onClick = { dialogueNouveauDossier = true }) {
                    Icon(Icons.Default.CreateNewFolder, contentDescription = null)
                    Spacer(Modifier.width(6.dp))
                    Text("Dossier")
                }
                OutlinedButton(onClick = { dialogueNouveauFichier = true }) {
                    Icon(Icons.Default.NoteAdd, contentDescription = null)
                    Spacer(Modifier.width(6.dp))
                    Text("Fichier")
                }
            }
            Spacer(Modifier.height(12.dp))

            erreur?.let {
                Text(it, color = MaterialTheme.colorScheme.error, style = MaterialTheme.typography.bodySmall)
                Spacer(Modifier.height(8.dp))
            }

            if (contenu.isEmpty()) {
                Text("Dossier vide.", style = MaterialTheme.typography.bodyMedium)
            }

            LazyColumn(modifier = Modifier.weight(1f)) {
                items(contenu) { element ->
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(vertical = 10.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Icon(
                            if (element.estDossier) Icons.Default.Folder else Icons.Default.InsertDriveFile,
                            contentDescription = null
                        )
                        Spacer(Modifier.width(12.dp))
                        Column(modifier = Modifier.weight(1f)) {
                            Text(
                                element.nom,
                                style = MaterialTheme.typography.bodyLarge
                            )
                            if (!element.estDossier) {
                                Text(
                                    "${element.tailleOctets / 1024} Ko",
                                    style = MaterialTheme.typography.bodySmall
                                )
                            }
                        }
                        if (element.estDossier) {
                            IconButton(onClick = {
                                pile = pile + Triple(element.uri, element.nom, courant.first)
                            }) {
                                Icon(Icons.Default.ChevronRight, contentDescription = "Ouvrir")
                            }
                        }
                        IconButton(onClick = { elementARenommer = element }) {
                            Icon(Icons.Default.Edit, contentDescription = "Renommer")
                        }
                        IconButton(onClick = { elementASupprimer = element }) {
                            Icon(Icons.Default.Delete, contentDescription = "Supprimer")
                        }
                    }
                    HorizontalDivider()
                }
            }
        }

        if (dialogueNouveauDossier) {
            DialogueTexte(
                titre = "Nouveau dossier",
                label = "Nom du dossier",
                onValider = { nom ->
                    erreur = if (repo.creerSousDossier(courant.first, nom)) null else "Impossible de créer ce dossier."
                    rafraichirContenu()
                    dialogueNouveauDossier = false
                },
                onAnnuler = { dialogueNouveauDossier = false }
            )
        }
        if (dialogueNouveauFichier) {
            DialogueTexte(
                titre = "Nouveau fichier",
                label = "Nom du fichier (avec extension)",
                onValider = { nom ->
                    erreur = if (repo.creerFichier(courant.first, nom)) null else "Impossible de créer ce fichier."
                    rafraichirContenu()
                    dialogueNouveauFichier = false
                },
                onAnnuler = { dialogueNouveauFichier = false }
            )
        }
        elementARenommer?.let { element ->
            DialogueTexte(
                titre = "Renommer",
                label = "Nouveau nom",
                valeurInitiale = element.nom,
                onValider = { nouveauNom ->
                    erreur = if (repo.renommer(element.uri, nouveauNom)) null else "Impossible de renommer."
                    rafraichirContenu()
                    elementARenommer = null
                },
                onAnnuler = { elementARenommer = null }
            )
        }
        elementASupprimer?.let { element ->
            AlertDialog(
                onDismissRequest = { elementASupprimer = null },
                title = { Text("Supprimer « ${element.nom} » ?") },
                text = { Text(if (element.estDossier) "Le dossier et tout son contenu seront supprimés." else "Le fichier sera supprimé définitivement.") },
                confirmButton = {
                    TextButton(onClick = {
                        erreur = if (repo.supprimer(element.uri)) null else "Impossible de supprimer."
                        rafraichirContenu()
                        elementASupprimer = null
                    }) { Text("Supprimer") }
                },
                dismissButton = {
                    TextButton(onClick = { elementASupprimer = null }) { Text("Annuler") }
                }
            )
        }
        return
    }

    // --- Vue racine : liste des dossiers designes ---
    Column(modifier = Modifier.fillMaxSize().padding(20.dp)) {
        Text("Dossiers désignés", style = MaterialTheme.typography.headlineSmall)
        Spacer(Modifier.height(4.dp))
        Text(
            "Clovis ne peut lire ou modifier que les dossiers que tu choisis explicitement ici.",
            style = MaterialTheme.typography.bodyMedium
        )
        Spacer(Modifier.height(16.dp))

        Button(onClick = { selecteurDossier.launch(null) }, modifier = Modifier.fillMaxWidth()) {
            Icon(Icons.Default.Add, contentDescription = null)
            Spacer(Modifier.width(6.dp))
            Text("Désigner un dossier")
        }
        Spacer(Modifier.height(16.dp))

        if (dossiersDesignes.isEmpty()) {
            Text("Aucun dossier désigné pour l'instant.", style = MaterialTheme.typography.bodyMedium)
        }

        LazyColumn {
            items(dossiersDesignes) { dossier: DossierDesigne ->
                Row(
                    modifier = Modifier.fillMaxWidth().padding(vertical = 10.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Icon(Icons.Default.Folder, contentDescription = null)
                    Spacer(Modifier.width(12.dp))
                    Text(
                        dossier.nom,
                        style = MaterialTheme.typography.bodyLarge,
                        modifier = Modifier.weight(1f)
                    )
                    IconButton(onClick = {
                        repo.retirerDossierDesigne(dossier.uri)
                        dossiersDesignes = repo.listerDossiersDesignes()
                    }) {
                        Icon(Icons.Default.Close, contentDescription = "Retirer")
                    }
                    IconButton(onClick = {
                        pile = listOf(Triple(dossier.uri, dossier.nom, null))
                    }) {
                        Icon(Icons.Default.ChevronRight, contentDescription = "Ouvrir")
                    }
                }
                HorizontalDivider()
            }
        }
    }
}

@Composable
private fun DialogueTexte(
    titre: String,
    label: String,
    valeurInitiale: String = "",
    onValider: (String) -> Unit,
    onAnnuler: () -> Unit
) {
    var valeur by remember { mutableStateOf(valeurInitiale) }
    AlertDialog(
        onDismissRequest = onAnnuler,
        title = { Text(titre) },
        text = {
            OutlinedTextField(
                value = valeur,
                onValueChange = { valeur = it },
                label = { Text(label) },
                modifier = Modifier.fillMaxWidth()
            )
        },
        confirmButton = {
            TextButton(onClick = { if (valeur.isNotBlank()) onValider(valeur) }, enabled = valeur.isNotBlank()) {
                Text("OK")
            }
        },
        dismissButton = {
            TextButton(onClick = onAnnuler) { Text("Annuler") }
        }
    )
}
